annif/backend/stwfsa.py

```diff
@@ -0,0 +1,121 @@
+_KEY_SIMPLE_ENGLISH_PLURAL_RULES = 'simple_english_plural_rules'
+_KEY_INPUT_LIMIT = 'input_limit'
+
+
+class StwfsaBackend(backend.AnnifBackend):
+
+    name = "stwfsa"
+    needs_subject_index = True
+
+    STWFSA_PARAMETERS = {
+        _KEY_CONCEPT_TYPE_URI: str,
+        _KEY_SUBTHESAURUS_TYPE_URI: str,
+        _KEY_THESAURUS_RELATION_TYPE_URI: str,
+        _KEY_THESAURUS_RELATION_IS_SPECIALISATION: boolean,
+        _KEY_REMOVE_DEPRECATED: boolean,
+        _KEY_HANDLE_TITLE_CASE: boolean,
+        _KEY_EXTRACT_UPPER_CASE_FROM_BRACES: boolean,
+        _KEY_EXTRACT_ANY_CASE_FROM_BRACES: boolean,
+        _KEY_EXPAND_AMPERSAND_WITH_SPACES: boolean,
+        _KEY_EXPAND_ABBREVIATION_WITH_PUNCTUATION: boolean,
+        _KEY_SIMPLE_ENGLISH_PLURAL_RULES: boolean,
+        _KEY_INPUT_LIMIT: int,
+    }
+
+    DEFAULT_PARAMETERS = {
+        _KEY_SUBTHESAURUS_TYPE_URI: '',
+        _KEY_THESAURUS_RELATION_TYPE_URI: '',
+        _KEY_THESAURUS_RELATION_IS_SPECIALISATION: False,
+        _KEY_REMOVE_DEPRECATED: True,
+        _KEY_HANDLE_TITLE_CASE: True,
+        _KEY_EXTRACT_UPPER_CASE_FROM_BRACES: True,
+        _KEY_EXTRACT_ANY_CASE_FROM_BRACES: False,
+        _KEY_EXPAND_AMPERSAND_WITH_SPACES: True,
+        _KEY_EXPAND_ABBREVIATION_WITH_PUNCTUATION: True,
+        _KEY_SIMPLE_ENGLISH_PLURAL_RULES: False,
+        _KEY_INPUT_LIMIT: 0,
+    }
+
+    MODEL_FILE = 'stwfsa_predictor.zip'
+
+    _model = None
+
+    def initialize(self):
+        if self._model is None:
+            path = os.path.join(self.datadir, self.MODEL_FILE)
+            self.debug(f'Loading STWFSA model from {path}.')
+            if os.path.exists(path):
+                self._model = StwfsapyPredictor.load(path)
+                self.debug('Loaded model.')
+            else:
+                raise NotInitializedException(
+                    f'Model not found at {path}',
+                    backend_id=self.backend_id)
+
+    def _train(self, corpus, params):
+        if corpus == 'cached':
+            raise NotSupportedException(
+                'Training stwfsa project from cached data not supported.')
+        if corpus.is_empty():
+            raise NotSupportedException(
+                'Cannot train stwfsa project with no documents.')
+        self.debug("Transforming training data.")
+        X = []
+        y = []
+        for doc in corpus.documents:
+            X.append(doc.text)
+            y.append(doc.uris)
+        graph = self.project.vocab.as_graph()
+        new_params = {
+                key: self.STWFSA_PARAMETERS[key](val)
+                for key, val
+                in params.items()
+                if key in self.STWFSA_PARAMETERS
+            }
+        new_params.pop(_KEY_INPUT_LIMIT)
+        p = StwfsapyPredictor(
+            graph=graph,
+            langs=frozenset([params['language']]),
+            **new_params)
+        p.fit(X, y)
+        self._model = p
+        p.store(os.path.join(self.datadir, self.MODEL_FILE))
```

Comment:
You could use `annif.util.atomic_save` here (like many other backends do) but no big deal if you don't. The idea of `atomic_save` is to first save the model (or whatever) into a temporary file, then rename it to the final name. If the program crashes during the save, it won't leave behind a broken model file.